firewood/src/merkle/node/branch.rs

```diff
@@ -232,13 +232,98 @@ impl Storable for BranchNode {
     }
 
     fn deserialize<T: crate::shale::CachedStore>(
-        _addr: usize,
-        _mem: &T,
-    ) -> Result<Self, crate::shale::ShaleError>
-    where
-        Self: Sized,
-    {
-        todo!()
+        mut addr: usize,
+        mem: &T,
+    ) -> Result<Self, crate::shale::ShaleError> {
+        const DATA_LEN_SIZE: usize = size_of::<DataLen>();
+        const BRANCH_HEADER_SIZE: u64 =
+            BranchNode::MAX_CHILDREN as u64 * DiskAddress::MSIZE + DATA_LEN_SIZE as u64;
+
+        let node_raw =
+            mem.get_view(addr, BRANCH_HEADER_SIZE)
+                .ok_or(ShaleError::InvalidCacheView {
+                    offset: addr,
+                    size: BRANCH_HEADER_SIZE,
+                })?;
+
+        addr += BRANCH_HEADER_SIZE as usize;
+
+        let mut cursor = Cursor::new(node_raw.as_deref());
+        let mut children = [None; BranchNode::MAX_CHILDREN];
+        let mut buf = [0u8; DiskAddress::MSIZE as usize];
+
+        for child in &mut children {
+            cursor.read_exact(&mut buf)?;
+            *child = Some(usize::from_le_bytes(buf))
+                .filter(|addr| *addr != 0)
+                .map(DiskAddress::from);
+        }
+
+        let raw_len = {
+            let mut buf = [0; DATA_LEN_SIZE];
+            cursor.read_exact(&mut buf)?;
```

Comment:
nit (DRY): identical code could be refactored into a macro, maybe:
```
macro_rules! read_data {
    ($cursor:expr, $type:ty) => {{
        let mut buf = [0u8; std::mem::size_of::<$type>()];
        $cursor.read_exact(&mut buf)?;
        <$type>::from_le_bytes(buf)
    }};
}
```
Then, at the callsite, you can do all the manipulation:
```
            *child = Some(read_data!(cursor, usize))
                .filter(|addr| *addr != 0)
                .map(DiskAddress::from);
```

Reply:
#396